Bluetooth: Log refactor [patch 2/??]

theob-pro opened the pull request:
We are trying to remove common/log.h in favor of standard Zephyr log functions. This PR a part of this effort.

In this PR:
- Create `common/bt_to_str` module for the Bluetooth string formatting functions that were in `common/log`.

See commit messages for details.

This PR will be merged after https://github.com/zephyrproject-rtos/zephyr/pull/51049.

theob-pro committed:
Bluetooth: Move out string related function of `common/log.h`

Functions related to string manipulation that were defined in
`common/log.h` has been moved to the `common/bt_str.h` file and their
implementation in `common/bt_str.c`.

Files that were using those functions has been updated consequently.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

theob-pro committed:
Bluetooth: Consolidate duplicated code

The file `tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c` was using
functions from `tests/bluetooth/host/host_mocks/print_utils.c` which were
exactly the same as the one in `common/log.h`.

The code was duplicated because the test author wanted to include
`common/log.h` to get the stringifying functions, but could not due to
unwanted side-effects.

See the comment from the unit test author here:
https://github.com/zephyrproject-rtos/zephyr/pull/48676#discussion_r973672151

The new `common/bt_str.h` does not have side-effects, so it is
suitable for unit tests.

The `snprintk` function has been redefined inside the `main.c` to use the
libc because of the kernel being not compiled entirely.

Also, the `CONFIG_LOG`, `CONFIG_BT_DEBUG_LOG` and
`CONFIG_TEST_LOGGING_DEFAULTS` have been disabled in the Kconfig files
of each `keys` tests. That so logs are not compiled and so, there is
no need to link the `printk` functions that are used by the logs.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>
Co-authored-by: Aleksander Wasaznik <aleksander.wasaznik@nordicsemi.no>

Thalley commented on subsys/bluetooth/common/bt_to_str.h:
```diff
@@ -0,0 +1,22 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/uuid.h>
+
+/* NOTE: These helper functions always encodes into the same buffer storage.
+ * It is the responsibility of the user of this function to copy the information
+ * in this string if needed.
+ *
+ * NOTE: These functions are not thread-safe!
+ */
+const char *bt_hex_real(const void *buf, size_t len);
+const char *bt_addr_str_real(const bt_addr_t *addr);
+const char *bt_addr_le_str_real(const bt_addr_le_t *addr);
+const char *bt_uuid_str_real(const struct bt_uuid *uuid);
+
+#define bt_hex(buf, len) bt_hex_real(buf, len)
+#define bt_addr_str(addr) bt_addr_str_real(addr)
+#define bt_addr_le_str(addr) bt_addr_le_str_real(addr)
+#define bt_uuid_str(uuid) bt_uuid_str_real(uuid)
```
I've always wondered: Is there a good reason why these are not available in the public API? 

Converting an addr to a string representation is something that we often do in samples an applications.

theob-pro replied:
Good point, I don't know either. We can consider doing that but in a separate PR.

alwa-nordic replied:
I think a good reason is that they are unsafe in concurrent code, and easy to misuse in general due to the static variable inside them. I have an idea for improved versions of these macros, which would allocate space on the stack before calling the real function. I think this version is safer and could be made public.

```c
/** Like bt_addr_le_to_str, but returns the pointer `out` instead of the write-length. */
char *bt_addr_le_str_ret(const bt_addr_le_t *addr, char *out)
{
	bt_addr_le_to_str(addr, out, BT_ADDR_LE_STR_LEN);
	return out;
}

/** The scope/lifetime of the returned string tied to the enclosing block.
  *
  * It can be safely used multiple times in a single statement and is thread-safe.
  *
  * `foo(bt_addr_le_str(addr))` equivalent with `char x[BT_ADDR_LE_STR_LEN];
  * foo(bt_addr_le_str_ret(addr, &x));`, but `x` is anonymous, so it cannot crash
  * with or shadow other definitions.
  */
#define bt_addr_le_str(addr) bt_addr_le_str_ret(addr, &(char[BT_ADDR_LE_STR_LEN]){})
```

Thalley commented on include/zephyr/bluetooth/addr.h:
```diff
@@ -213,7 +214,7 @@ static inline bool bt_addr_le_is_identity(const bt_addr_le_t *addr)
 static inline int bt_addr_to_str(const bt_addr_t *addr, char *str, size_t len)
 {
-	return snprintk(str, len, "%02X:%02X:%02X:%02X:%02X:%02X",
+	return snprintf(str, len, "%02X:%02X:%02X:%02X:%02X:%02X",
```
You mention the reason for changing this from `snprintk` to `snprintf` was due to linker issues. Can you elaborate?

jori-nordic replied:
we didn't want to pull in `log_minimal.c`, the original test had a copy-pasted `z_log_minimal_printk()` to avoid this.
We also had issues with `snprintk` pulling in additional depencies, like `cbvprintf`.
So we just thought it'd be okay to change it to use the libc rather than the kernel facilities. 

Do you know if there are any situations where one should prefer using `snprintk` over `snprintf`? we didn't think of any but we don't know the kernel that much, if you have any we can find an alternative to this change.

Thalley replied:
I figured that in Zephyr context we would always use the `k` functions, to avoid any issues when changing stdlibs. 

If there are issues with `snprintk`, shouldn't they be solved, instead of just replacing them?

jhedberg replied:
One of the reasons for preferring snprintk in the past is that it has had a substantially smaller stack footprint.

jhedberg replied:
I also wanted to add that we should avoid linking two different formatters into the final image, if possible. I.e. either we should be using the printk formatter all across the board, or then use libc everywhere.

jori-nordic replied:
> One of the reasons for preferring snprintk in the past is that it has had a substantially smaller stack footprint.

I think this is not true anymore, since the stack size didn't increase that much (80 bytes) when we switched to libc.

theob-pro replied:
I think that fixing the `snprintk` thing would be out of the scope of this PR. I reverted the change.

Thalley commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -11,11 +11,25 @@
-#include "host_mocks/print_utils.h"
 #include "testing_common_defs.h"
+#include "common/bt_to_str.h"
+#include <stdio.h>
```
Don't we usually have the standard includes at the top?

theob-pro replied:
Fixed.

jhedberg replied:
Yes, we should be consistent about this, but note that this will likely change in the future. Header ordering was discussed in yesterday's dev meeting and will likely come to the architecture meeting next week or in the near future. 

See #41543 and https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes which is the style that may eventually be adopted by Zephyr.

Thalley commented on tests/bluetooth/host/keys/bt_keys_foreach_bond/prj.conf:
```diff
@@ -6,3 +6,6 @@ CONFIG_BT_MAX_PAIRED=4
 CONFIG_ASSERT=y
 CONFIG_ASSERT_LEVEL=2
 CONFIG_ASSERT_VERBOSE=y
+CONFIG_LOG=n
+CONFIG_BT_DEBUG_LOG=n
+CONFIG_TEST_LOGGING_DEFAULTS=n
```
You've added these a few places. I am not sure I understand the purpose of this, can you elaborate?

theob-pro replied:
Yes, I added that so the logs in files like `bluetooth/host/keys.c` that are used by the tests are not compiled and so there is no need to link the `printk` functions that are used by the logs.

Thalley replied:
Can I bother you to add that to the commit message? At least that is to me not an obvious change, nor is the reasoning part of commit message :)

theob-pro replied:
You're right, I'll do it.

Thalley reviewed (dismissed):
ACK for BLE Audio

alwa-nordic commented on tests/bluetooth/host/keys/bt_keys_get_addr/src/main.c:
```diff
@@ -4,18 +4,32 @@
 
 DEFINE_FFF_GLOBALS;
 
+/* Define snprintk to use libc since we are not compiling the whole kernel. */
+int snprintk(char *str, size_t size, const char *fmt, ...)
+{
+	va_list ap;
+	int ret;
+
+	va_start(ap, fmt);
+	ret = snprintf(str, size, fmt, ap);
+	va_end(ap);
+
+	return ret;
+}
+
```
I think this should go in `subsys/testsuite/ztest/src/ztest_mock.c`.

theob-pro committed:
Bluetooth: Move re-implementation of `snprintk`

Move the function in the `subsys/testsuite/ztest/src/ztest_mock.c` files.

This is motivated by the fact that there is others re-implementation of
`*printk` functions using libc counterparts in the `ztest_mock.c` file.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

carlescufi reviewed (changes requested):
Why this commit:
`Bluetooth: Move out string related function of common/log.h`
?

carlescufi commented on subsys/bluetooth/common/bt_to_str.c:
```diff
@@ -0,0 +1,63 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/types.h>
+
+const char *bt_hex_real(const void *buf, size_t len)
```
Do we still need this function at all? Don't we have this in the logging subsystem now?

Thalley replied:
I guess you are referring to `LOG_HEXDUMP_DBG`? 

I think the two are slightly different, as `LOG_HEXDUMP_DBG` doesn't support formatting, so if you want to log a value and an array, you can't with `LOG_HEXDUMP_DBG`, but you can do

```c
BT_DBG("val %u array val %s", val, bt_hex_real(array_val, sizeof(array_val)));
```

So IMO `LOG_HEXDUMP_DBG` doesn't replace `bt_hex_real`.

carlescufi replied:
That makes sense, thanks for the explanation @Thalley. We probably want to move this function to `subsys/logging` though. But we can leave that for later, for now no objections to keeping it here.

carlescufi commented:
> Why this commit: `Bluetooth: Move out string related function of common/log.h` ?

I see this is for reusing in the Unit Tests, so fine. Not a fan of the `bt_to_str` name but I can live with it. I would rather have it named "bt_str.{c,h}`. @jhedberg ?

jhedberg commented on subsys/bluetooth/common/bt_to_str.h:
```diff
@@ -15,8 +15,3 @@ const char *bt_hex_real(const void *buf, size_t len);
-#define bt_hex(buf, len) bt_hex_real(buf, len)
-#define bt_addr_str(addr) bt_addr_str_real(addr)
-#define bt_addr_le_str(addr) bt_addr_le_str_real(addr)
-#define bt_uuid_str(uuid) bt_uuid_str_real(uuid)
```
You should also rename the actual c-functions, i.e. remove the `_real` suffix - that has no meaning since we're not wrapping the functions anymore.

theob-pro replied:
Done.

Thalley commented on subsys/bluetooth/common/bt_to_str.h:
```diff
@@ -0,0 +1,17 @@
+ *
+ * NOTE: These functions are not thread-safe!
+ */
+const char *bt_hex_real(const void *buf, size_t len);
+const char *bt_addr_str_real(const bt_addr_t *addr);
+const char *bt_addr_le_str_real(const bt_addr_le_t *addr);
+const char *bt_uuid_str_real(const struct bt_uuid *uuid);
```
I think it was requested to only keep the non-`real` functions, so these should have been
```suggestion
const char *bt_hex(const void *buf, size_t len);
const char *bt_addr_str(const bt_addr_t *addr);
const char *bt_addr_le_str(const bt_addr_le_t *addr);
const char *bt_uuid_str(const struct bt_uuid *uuid);
```

theob-pro committed:
Bluetooth: Remove unneeded macros

Remove the `bt_hex`, `bt_addr_str`, `bt_addr_le_str` and `bt_uuid_str`
macros from `bt_str.h` as they were just aliases for the `*_real` functions
in the same file.

The functions has been renamed without the `_real` suffix.

Some files were using the functions and not the macros, they have been
updated to use the new name of the functions.

Signed-off-by: Théo Battrel <theo.battrel@nordicsemi.no>

Thalley commented:
@theob-pro Please rebase to solve merge conflicts :)

jori-nordic commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -25,6 +25,7 @@
 #define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
 #define LOG_MODULE_NAME bt_driver
 #include "common/log.h"
+#include "common/bt_str.h"
```
can't comment on the commit msg, but you forgot to update the names (bt_to_str -> bt_str) there.
